File: Problem-061/solution.py
def cyclical_figurate_nums(n):

    formulas = {
        3: lambda x: x * (x + 1) // 2,
        4: lambda x: x * x,
        5: lambda x: x * (3 * x - 1) // 2,
        6: lambda x: x * (2 * x - 1),
        7: lambda x: x * (5 * x - 3) // 2,
        8: lambda x: x * (3 * x - 2)
    }

    numbers = {}

    # Generate figurate numbers
    for type_ in range(3, n + 3):

        numbers[type_] = []

        x = 1

        while True:

            value = formulas[type_](x)

            if value >= 10000:
                break

            if value >= 1000 and value % 100 >= 10:
                numbers[type_].append(value)

            x += 1

    # prefix -> possible numbers
    prefix_map = {}

    for type_ in range(3, n + 3):

        for value in numbers[type_]:

            prefix = value // 100

            if prefix not in prefix_map:
                prefix_map[prefix] = []

            prefix_map[prefix].append({
                "value": value,
                "type": type_
            })

    total = 0

    # Store already-found cycles
    found = set()

    def search(chain, used_types, used_numbers):

        nonlocal total

        # We have n numbers
        if len(chain) == n:

            last = chain[-1]
            first = chain[0]

            # Check cyclic condition
            if last % 100 != first // 100:
                return

            # Prevent counting rotations of same cycle
            key = tuple(sorted(chain))

            if key in found:
                return

            found.add(key)

            total += sum(chain)

            return

        last = chain[-1]

        # Last two digits determine next prefix
        suffix = last % 100

        candidates = prefix_map.get(suffix)

        if not candidates:
            return

        for candidate in candidates:

            candidate_value = candidate["value"]
            candidate_type = candidate["type"]

            # Each figurate type only once
            if candidate_type in used_types:
                continue

            # Each number only once
            if candidate_value in used_numbers:
                continue

            used_types.add(candidate_type)
            used_numbers.add(candidate_value)
            chain.append(candidate_value)

            search(
                chain,
                used_types,
                used_numbers
            )

            chain.pop()
            used_types.remove(candidate_type)
            used_numbers.remove(candidate_value)

    # Start search
    for type_ in range(3, n + 3):

        for value in numbers[type_]:

            search(
                [value],
                {type_},
                {value}
            )

    return total

File: Problem-061/solution.js
function cyclicalFigurateNums(n) {

    const formulas = {
        3: x => x * (x + 1) / 2,
        4: x => x * x,
        5: x => x * (3 * x - 1) / 2,
        6: x => x * (2 * x - 1),
        7: x => x * (5 * x - 3) / 2,
        8: x => x * (3 * x - 2)
    };
    const numbers = {};

    for (let type = 3; type <= n + 2; type++) {

        numbers[type] = [];

        for (let x = 1; ; x++) {

            const value = formulas[type](x);

            if (value >= 10000) {
                break;
            }
            if (value >= 1000 && value % 100 >= 10) {
                numbers[type].push(value);
            }
        }
    }


    const prefixMap = new Map();

    for (let type = 3; type <= n + 2; type++) {

        for (const value of numbers[type]) {

            const prefix = Math.floor(value / 100);

            if (!prefixMap.has(prefix)) {
                prefixMap.set(prefix, []);
            }

            prefixMap.get(prefix).push({
                value: value,
                type: type
            });
        }
    }

    let total = 0;

    const found = new Set();

    function search(chain, usedTypes, usedNumbers) {

        if (chain.length === n) {
            const last = chain[chain.length - 1];
            const first = chain[0];

            if (last % 100 !== Math.floor(first / 100)) {
                return;
            }
            const key = [...chain].sort((a, b) => a - b).join(",");

            if (found.has(key)) {
                return;
            }

            found.add(key);

            let sum = 0;

            for (const value of chain) {
                sum += value;
            }

            total += sum;

            return;
        }

        const last = chain[chain.length - 1];

        const suffix = last % 100;

        const candidates = prefixMap.get(suffix);

        if (!candidates) {
            return;
        }

        for (const candidate of candidates) {
            if (usedTypes.has(candidate.type)) {
                continue;
            }

            if (usedNumbers.has(candidate.value)) {
                continue;
            }

            usedTypes.add(candidate.type);
            usedNumbers.add(candidate.value);
            chain.push(candidate.value);

            search(
                chain,
                usedTypes,
                usedNumbers
            );

            chain.pop();
            usedTypes.delete(candidate.type);
            usedNumbers.delete(candidate.value);
        }
    }

    for (let type = 3; type <= n + 2; type++) {

        for (const value of numbers[type]) {

            search(
                [value],
                new Set([type]),
                new Set([value])
            );
        }
    }

    return total;
}
